src/WebJobs.Script.Grpc/Proto/FunctionRpc.proto

```diff
@@ -0,0 +1,255 @@
+// Worker self terminates ..no response on this
+message WorkerTerminate {
+  google.protobuf.Duration grace_period = 1;
+}
+
+message FileChangeEventRequest {
+  // https://msdn.microsoft.com/en-us/library/t6xf43e0(v=vs.110).aspx
+  enum Type {
+	Unknown = 0;
+    Created = 1;
+    Deleted = 2;
+    Changed = 4;
+    Renamed = 8;
+    All = 15;
+  }
+
+  Type type = 1;
+  string full_path = 2;
+  string name = 3;
+}
+
+message WorkerActionResponse {
+  enum Action {
+    Restart = 0;
+    Reload = 1;
+  }
+  
+  Action action = 1;
+  string reason = 2;
+}
+
+message WorkerStatusRequest{
+}
+
+message WorkerStatusResponse {
+}
+
+message FunctionLoadRequest {
+  // unique function identifier (avoid name collisions, facilitate reload case)
+  string function_id = 1;
+  RpcFunctionMetadata metadata = 2;
+}
+
+message FunctionLoadResponse {
+  string function_id = 1;
+  StatusResult result = 2;
+  // TODO: return type expected?
+}
+
+message RpcFunctionMetadata {
+  // TODO: do we want the host's name - the language worker might do a better job of assignment than the host
+  string name = 4;
+
+  string directory = 1;
+  string script_file = 2;
+  string entry_point = 3;
+
+  map<string, BindingInfo> bindings = 6;
+
+  // not adding disabled or excluded as those (currently) are only relevant to host
+}
+
+message InvocationRequest {
+  string invocation_id = 1;
+  string function_id = 2;
+  repeated ParameterBinding input_data = 3;
+  map<string, TypedData> trigger_metadata = 4;
+}
+
+message InvocationCancel {
+  string invocation_id = 2;
+  google.protobuf.Duration grace_period = 1; // could also use absolute time
+}
+
+message InvocationResponse {
+  string invocation_id = 1;
+  repeated ParameterBinding output_data = 2;
+  TypedData return_value = 4;
+  StatusResult result = 3;
+}
+
+message TypedData {
+  oneof data {
+    string string = 1;
+    string json = 2;
+    bytes bytes = 3;
+    bytes stream = 4;
+    RpcHttp http = 5;
+	sint64 int = 6;
+    double double = 7;
+  }
+}
+
+message ParameterBinding {
+  string name = 1;
+  TypedData data = 2;
+}
+
+message BindingInfo {
+  string type = 2;
+  Direction direction = 3;
+
+  enum Direction {
+    in = 0;
+	out = 1;
+	inout = 2;
+  }
+}
+
+message RpcLog {
+  // Matching ILogger semantics
+  // https://github.com/aspnet/Logging/blob/9506ccc3f3491488fe88010ef8b9eb64594abf95/src/Microsoft.Extensions.Logging/Logger.cs
+  enum Level {
+    Trace = 0;
+    Debug = 1;
+    Information = 2;
+    Warning = 3;
+    Error = 4;
+    Critical = 5;
+    None = 6;
+  }
+
+  string invocation_id = 1;
+  string category = 2;
+  Level level = 3;
+  string message = 4;
+  string event_id = 5;
+  RpcException exception = 6;
+
+  // json serialized property bag, or could use a type scheme like map<string, TypedData>
+  string properties = 7;
+}
+
+message RpcException {
+  string source = 3;
+  string stack_trace = 1;
+  string message = 2;
+}
+
+// TODO - solidify this
```

Comment:
Why do we have Http specific message, but not Blob/Queue/other specifics? Ie, what makes Http special?

Reply:
They all suffer from the rich types problem, which is that json does not provide enough type information. The case I fall back to is distinguishing between a byte array and a base 64 serialized string when json serialized.